tests/test_base_device.py

```diff
@@ -1114,3 +1174,171 @@ def qiskit_circuit(x):
+
+    @pytest.mark.parametrize(
+        "observable",
+        [
+            lambda: [qml.expval(qml.Hadamard(0)), qml.expval(qml.Hadamard(0))],
+            lambda: [
+                qml.var(qml.Hadamard(0)),
+                qml.var(qml.Hadamard(0)),
+            ],  # Note that Sum would fail
+            lambda: [
+                qml.expval(qml.X(0)),
+                qml.expval(qml.Y(1)),
+                qml.expval(0.5 * qml.Y(1)),
+                qml.expval(qml.Z(0) @ qml.Z(1)),
+                qml.expval(qml.X(0) @ qml.Z(1) + 0.5 * qml.Y(1) + qml.Z(0)),
+                qml.expval(
+                    qml.ops.LinearCombination(
+                        [0.35, 0.46], [qml.X(0) @ qml.Z(1), qml.Z(0) @ qml.X(2)]
+                    )
+                ),
+                qml.expval(
+                    qml.ops.LinearCombination(
+                        [1.0, 2.0, 3.0], [qml.X(0), qml.X(1), qml.Z(0)], grouping_type="qwc"
+                    )
+                ),
+            ],
+            lambda: [
+                qml.expval(
+                    qml.Hamiltonian([0.35, 0.46], [qml.X(0) @ qml.Z(1), qml.Z(0) @ qml.Y(2)])
+                )
+            ],
+            lambda: [qml.expval(qml.X(0) @ qml.Z(1) + qml.Z(0))],
+            lambda: [
+                qml.var(qml.X(1) + qml.Z(0))
+            ],  # This one is not non-commuting, but nice for comparison
+            pytest.param(
+                [qml.var(qml.X(0) + qml.Z(0))],
+                marks=pytest.mark.xfail(reason="Qiskit itself is bugged when given Sum"),
+            ),
+            lambda: [
+                qml.expval(qml.Hadamard(0)),
+                qml.expval(qml.Hadamard(1)),
+                qml.expval(qml.Hadamard(0) @ qml.Hadamard(1)),
+                qml.expval(
+                    qml.Hadamard(0) @ qml.Hadamard(1) + 0.5 * qml.Hadamard(1) + qml.Hadamard(0)
+                ),
+            ],
+        ],
+    )
+    def test_observables_that_need_split_non_commuting(self, observable):
+        """Tests that observables that have non-commuting measurements are
+        processed correctly when executed by the Estimator or, in the case of
+        qml.Hadamard, executed by the Sampler via expval() or var"""
+        qiskit_dev = QiskitDevice2(wires=3, backend=backend, shots=30000)
+
+        @qml.qnode(qiskit_dev)
+        def qiskit_circuit():
+            qml.RX(np.pi / 3, 0)
+            qml.RZ(np.pi / 3, 0)
+            return observable()
+
+        dev = qml.device("default.qubit", wires=3, shots=30000)
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.RX(np.pi / 3, 0)
+            qml.RZ(np.pi / 3, 0)
+            return observable()
+
+        qiskit_res = qiskit_circuit()
+        res = circuit()
+
+        assert np.allclose(res, qiskit_res, atol=0.05)
+
+    @pytest.mark.parametrize(
+        "observable",
+        [
+            lambda: [qml.counts(qml.X(0) + qml.Y(0)), qml.counts(qml.X(0))],
```

Comment:
Instead of using `lambda`, you can use `qml.apply(measurement)` to queue the measurement.

Reply:
Ohh didn't know that. Thanks @obliviateandsurrender!

Reply:
Seems like qml.apply is bugged for operators like `[qml.counts(qml.X(0) + qml.Y(0)), qml.counts(qml.X(0))]`

```
dev = qml.device("default.qubit", wires=2, shots=1024)

op = qml.sample(qml.X(0) + qml.Y(0)), qml.sample(qml.X(0))
@qml.qnode(dev)
def circuit():
    qml.RX(np.pi / 3, 0)
    qml.RZ(np.pi / 3, 0)
    return qml.apply(op)

>>> circuit()
()
```

Reply:
For now I will stick with lambda since it works. If this is a bug I can put a TODO and fix it for later, but I don't think it's that big of a deal.

Reply:
Seems like it is not designed to work like this, I will use lambda.